Cannot convert frame to type uintptr

**Describe the bug**
A clear and concise description of what the bug is.

On my CI server, i try to run a SAST with gosec, but I get a compile-time error, saying
```
# go.elastic.co/apm
../go.elastic.co/apm/error.go:361:19: cannot convert frame (type "github.com/pkg/errors".Frame) to type uintptr
```

**To Reproduce**
Steps to reproduce the behavior:
1. Run `docker run --env SAST_CONFIDENCE_LEVEL="${SAST_CONFIDENCE_LEVEL:-3}" --volume "$PWD:/code" --volume /var/run/docker.sock:/var/run/docker.sock "registry.gitlab.com/gitlab-org/security-products/sast:$SP_VERSION" /app/bin/run /code` on my codebase
(The only location where I use APM is in [`server.go`](https://gitlab.com/iwritive/potassium/blob/develop/server.go#L37-39) as a gRPC unary interceptor)

As in [https://gitlab.com/iwritive/potassium/-/jobs/141919551](https://gitlab.com/iwritive/potassium/-/jobs/141919551)

**Expected behavior**
A clear and concise description of what you expected to happen.

I don't get why there is a compile-time error, but to me it doesn't seem like a problem that can be solved by a re-compile (and it doesn't).

Maybe this is a problem with cross-platform compilation?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,8 +1,11 @@
 # Changelog
 
-## [Unreleased](https://github.com/elastic/apm-agent-go/compare/v1.1.2...master)
+## [Unreleased](https://github.com/elastic/apm-agent-go/compare/v1.1.3...master)
+
+## [v1.1.3](https://github.com/elastic/apm-agent-go/releases/tag/v1.1.3)
 
  - Remove the `agent.*` metrics (#407)
+ - Add support for new github.com/pkg/errors.Frame type (#409)
 
 ## [v1.1.2](https://github.com/elastic/apm-agent-go/releases/tag/v1.1.2)
 

--- a/error.go
+++ b/error.go
@@ -6,12 +6,21 @@ import (
 	"net"
 	"os"
 	"reflect"
+	"runtime"
 	"syscall"
 	"time"
+	"unsafe"
 
 	"github.com/pkg/errors"
 
 	"go.elastic.co/apm/stacktrace"
+)
+
+var (
+	uintptrType             = reflect.TypeOf(uintptr(0))
+	runtimeFrameType        = reflect.TypeOf(runtime.Frame{})
+	errorsStackTraceUintptr = uintptrType.ConvertibleTo(reflect.TypeOf(*new(errors.Frame)))
+	errorsStackTraceFrame   = reflect.TypeOf(*new(errors.Frame)).ConvertibleTo(runtimeFrameType)
 )
 
 // Recovered creates an Error with t.NewError(err), where
@@ -355,12 +364,26 @@ func initStacktrace(e *Error, err error) {
 	case internalStackTracer:
 		e.stacktrace = append(e.stacktrace[:0], stackTracer.StackTrace()...)
 	case errorsStackTracer:
+		// github.com/pkg/errors 0.8.x and earlier represent
+		// stack frames as uintptr; 0.9.0 and later represent
+		// them as runtime.Frames.
+		//
+		// TODO(axw) drop support for older github.com/pkg/errors
+		// versions when we release go.elastic.co/apm v2.0.0.
 		stackTrace := stackTracer.StackTrace()
-		pc := make([]uintptr, len(stackTrace))
-		for i, frame := range stackTrace {
-			pc[i] = uintptr(frame)
+		if errorsStackTraceUintptr {
+			pc := make([]uintptr, len(stackTrace))
+			for i, frame := range stackTrace {
+				pc[i] = *(*uintptr)(unsafe.Pointer(&frame))
+			}
+			e.stacktrace = stacktrace.AppendCallerFrames(e.stacktrace[:0], pc, -1)
+		} else if errorsStackTraceFrame {
+			e.stacktrace = e.stacktrace[:0]
+			for _, frame := range stackTrace {
+				rf := (*runtime.Frame)(unsafe.Pointer(&frame))
+				e.stacktrace = append(e.stacktrace, stacktrace.RuntimeFrame(*rf))
+			}
 		}
-		e.stacktrace = stacktrace.AppendCallerFrames(e.stacktrace[:0], pc, -1)
 	}
 }
 

--- a/error_test.go
+++ b/error_test.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"path/filepath"
+	"reflect"
 	"runtime"
 	"testing"
 
@@ -171,9 +172,33 @@ func (e *errorsStackTracer) StackTrace() errors.StackTrace {
 func newErrorsStackTrace(skip, n int) errors.StackTrace {
 	callers := make([]uintptr, 2)
 	callers = callers[:runtime.Callers(1, callers)]
-	frames := make([]errors.Frame, len(callers))
-	for i, pc := range callers {
-		frames[i] = errors.Frame(pc)
+
+	var (
+		uintptrType      = reflect.TypeOf(uintptr(0))
+		errorsFrameType  = reflect.TypeOf(*new(errors.Frame))
+		runtimeFrameType = reflect.TypeOf(runtime.Frame{})
+	)
+
+	var frames []errors.Frame
+	switch {
+	case errorsFrameType.ConvertibleTo(uintptrType):
+		frames = make([]errors.Frame, len(callers))
+		for i, pc := range callers {
+			reflect.ValueOf(&frames[i]).Elem().Set(reflect.ValueOf(pc).Convert(errorsFrameType))
+		}
+	case errorsFrameType.ConvertibleTo(runtimeFrameType):
+		fs := runtime.CallersFrames(callers)
+		for {
+			var frame errors.Frame
+			runtimeFrame, more := fs.Next()
+			reflect.ValueOf(&frame).Elem().Set(reflect.ValueOf(runtimeFrame).Convert(errorsFrameType))
+			frames = append(frames, frame)
+			if !more {
+				break
+			}
+		}
+	default:
+		panic(fmt.Errorf("unhandled errors.Frame type %s", errorsFrameType))
 	}
 	return errors.StackTrace(frames)
 }

--- a/version.go
+++ b/version.go
@@ -2,5 +2,5 @@ package apm
 
 const (
 	// AgentVersion is the Elastic APM Go Agent version.
-	AgentVersion = "1.1.2"
+	AgentVersion = "1.1.3"
 )

